web/web.go

```diff
@@ -0,0 +1,773 @@
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	var dataSourceName string
+	switch config.Driver {
+	case "postgres":
+		dataSourceName = fmt.Sprintf("host=%s port=%s user=%s password=%s dbname=%s sslmode=disable", config.Host, config.Port, config.User, config.Password, config.Database)
+	case "mysql":
+		dataSourceName = fmt.Sprintf("%s:%s@tcp(%s:%s)/%s", config.User, config.Password, config.Host, config.Port, config.Database)
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", config.Driver), http.StatusBadRequest)
+		return
+	}
+	sourceDB, err := sql.Open(config.Driver, dataSourceName)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("SQL connection error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	// Open doesn't open a connection. Validate database connection.
+	err = sourceDB.Ping()
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Connection Error: %v. Check Configuration again.", err), http.StatusInternalServerError)
+		return
+	}
+	app.sourceDB = sourceDB
+	app.dbName = config.Database
+	app.driver = config.Driver
+	app.sessionFile = ""
+	w.WriteHeader(http.StatusOK)
+}
+
+// convertSchemaSQL converts source database to Spanner when using
+// with postgres and mysql driver.
+func convertSchemaSQL(w http.ResponseWriter, r *http.Request) {
+	if app.sourceDB == nil || app.dbName == "" || app.driver == "" {
+		http.Error(w, fmt.Sprintf("Database is not configured or Database connection is lost. Please set configuration and connect to database."), http.StatusNotFound)
+		return
+	}
+	conv := internal.MakeConv()
+	var err error
+	switch app.driver {
+	case "mysql":
+		err = mysql.ProcessInfoSchema(conv, app.sourceDB, app.dbName)
+	case "postgres":
+		err = postgres.ProcessInfoSchema(conv, app.sourceDB)
+	default:
+		http.Error(w, fmt.Sprintf("Driver : '%s' is not supported", app.driver), http.StatusBadRequest)
+		return
+	}
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Schema Conversion Error : %v", err), http.StatusNotFound)
+		return
+	}
+	app.conv = conv
+	w.WriteHeader(http.StatusOK)
+	json.NewEncoder(w).Encode(conv)
+}
+
+// dumpConfig contains the parameters needed to run the tool using dump approach. It is
+// used to communicate via HTTP with the frontend.
+type dumpConfig struct {
+	Driver   string `json:"Driver"`
+	FilePath string `json:"Path"`
+}
+
+// convertSchemaDump converts schema from dump file to Spanner schema for
+// mysqldump and pg_dump driver.
+func convertSchemaDump(w http.ResponseWriter, r *http.Request) {
+	reqBody, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Body Read Error : %v", err), http.StatusInternalServerError)
+		return
+	}
+	var dc dumpConfig
+	err = json.Unmarshal(reqBody, &dc)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Request Body parse error : %v", err), http.StatusBadRequest)
+		return
+	}
+	f, err := os.Open(dc.FilePath)
```

Comment:
So when we open a dump file from the web UI, we look for the file on the file system that the server is running on.

If the web UI is running on a browser on the same machine as the server, then all will be good.

But if the web UI is running on a browser on a different machine, this this might be confusing.
For example, I might run HarbourBridge on my workstation, and then get the web UI up on a browser on my laptop. When I select a file, I might think it should be a file on my laptop?

I'm not sure we want to deal with all of the issues of browser file upload, but maybe we should document this somewhere?

Reply:
Thanks!
Closing thread.